Source/WebCore/platform/ScrollSnapAnimatorState.cpp

```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    }
+}
+
+void ScrollSnapAnimatorState::preserveCurrentTargetForAxis(ScrollEventAxis axis)
+{
+    auto snapID = activeSnapIDForAxis(axis);
+    auto snapOffsets = snapOffsetsForAxis(axis);
+    
+    auto found = std::find_if(snapOffsets.begin(), snapOffsets.end(), [snapID](SnapOffset<LayoutUnit> p) -> bool {
+        return p.snapTargetId == *snapID;
+    });
+    if (found != snapOffsets.end())
+        setActiveSnapIndexForAxis(axis, std::distance(snapOffsets.begin(), found));
+}
+
 bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const ScrollExtents& scrollExtents, float pageScale)
 {
+    // Check if we need to set the active target to a focused element
+    setFocusedElementForAxis(ScrollEventAxis::Vertical);
+    setFocusedElementForAxis(ScrollEventAxis::Horizontal);
+
     bool snapPointChanged = false;
-    // If we are already snapped in a particular axis, maintain that. Otherwise, snap to the nearest eligible snap point.
     auto activeHorizontalIndex = activeSnapIndexForAxis(ScrollEventAxis::Horizontal);
-    if (!activeHorizontalIndex || *activeHorizontalIndex >= snapOffsetsForAxis(ScrollEventAxis::Horizontal).size())
-        snapPointChanged |= setNearestScrollSnapIndexForAxisAndOffset(ScrollEventAxis::Horizontal, scrollOffset, scrollExtents, pageScale);
-
     auto activeVerticalIndex = activeSnapIndexForAxis(ScrollEventAxis::Vertical);
-    if (!activeVerticalIndex || *activeVerticalIndex >= snapOffsetsForAxis(ScrollEventAxis::Vertical).size())
+    auto activeHorizontalID = activeSnapIDForAxis(ScrollEventAxis::Horizontal);
+    auto activeVerticalID = activeSnapIDForAxis(ScrollEventAxis::Vertical);
+    auto snapOffsetsVertical = snapOffsetsForAxis(ScrollEventAxis::Vertical);
+    auto snapOffsetsHorizontal = snapOffsetsForAxis(ScrollEventAxis::Horizontal);
+    
+    // Check if we need to set the current indeces
+    if (!activeVerticalIndex || *activeVerticalIndex >= snapOffsetsForAxis(ScrollEventAxis::Vertical).size()) {
         snapPointChanged |= setNearestScrollSnapIndexForAxisAndOffset(ScrollEventAxis::Vertical, scrollOffset, scrollExtents, pageScale);
-
+        activeVerticalIndex = activeSnapIndexForAxis(ScrollEventAxis::Vertical);
+    }
+    if (!activeHorizontalIndex || *activeHorizontalIndex >= snapOffsetsForAxis(ScrollEventAxis::Horizontal).size()) {
+        snapPointChanged |= setNearestScrollSnapIndexForAxisAndOffset(ScrollEventAxis::Horizontal, scrollOffset, scrollExtents, pageScale);
+        activeHorizontalIndex = activeSnapIndexForAxis(ScrollEventAxis::Horizontal);
+    }
+    
+    // If we have an active target, see if we need to preserve it
+    if (activeHorizontalID && activeHorizontalIndex) {
+        if (*activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetId)
+            preserveCurrentTargetForAxis(ScrollEventAxis::Horizontal);
```

Comment:
Should this update `snapPointChanged` as well? I think it might need to so that the new snap index is sent to the Scrolling Tree.

Reply:
Fixed.